openmetadata-service/src/main/java/org/openmetadata/service/apps/bundles/changeEvent/AbstractEventConsumer.java

```diff
@@ -0,0 +1,247 @@
+    EventSubscriptionOffset eventSubscriptionOffset =
+        new EventSubscriptionOffset().withOffset(offset).withTimestamp(System.currentTimeMillis());
+    Entity.getCollectionDAO()
+        .eventSubscriptionDAO()
+        .upsertSubscriberOffset(
+            eventSubscription.getId().toString(),
+            OFFSET_EXTENSION,
+            "eventSubscriptionOffset",
+            JsonUtils.pojoToJson(eventSubscriptionOffset));
+
+    // Update the Job Data Map with the latest offset
+    jobExecutionContext.getJobDetail().getJobDataMap().put(ALERT_OFFSET_KEY, offset);
+  }
+
+  public synchronized void setErrorStatus(Long attemptTime, Integer statusCode, String reason) {
+    setStatus(FAILED, attemptTime, statusCode, reason, null);
+  }
+
+  public synchronized void setAwaitingRetry(Long attemptTime, int statusCode, String reason) {
+    setStatus(AWAITING_RETRY, attemptTime, statusCode, reason, attemptTime + currentBackoffTime);
+  }
+
+  public synchronized void setSuccessStatus(Long updateTime) {
+    SubscriptionStatus subStatus =
+        AlertUtil.buildSubscriptionStatus(
+            ACTIVE, updateTime, null, null, null, updateTime, updateTime);
+    eventSubscription.setStatusDetails(subStatus);
+  }
+
+  protected synchronized void setStatus(
+      SubscriptionStatus.Status status,
+      Long attemptTime,
+      Integer statusCode,
+      String reason,
+      Long timestamp) {
+    SubscriptionStatus subStatus =
+        AlertUtil.buildSubscriptionStatus(
+            status, null, attemptTime, statusCode, reason, timestamp, attemptTime);
+    eventSubscription.setStatusDetails(subStatus);
+  }
+
+  @Override
+  public List<ChangeEvent> pollEvents(long offset, long batchSize) {
+    // Read from Change Event Table
+    List<String> eventJson = Entity.getCollectionDAO().changeEventDAO().list(batchSize, offset);
+
+    List<ChangeEvent> changeEvents = new ArrayList<>();
+    for (String json : eventJson) {
+      ChangeEvent event = JsonUtils.readValue(json, ChangeEvent.class);
+      changeEvents.add(event);
+    }
+    return changeEvents;
+  }
+
+  @Override
+  public void execute(JobExecutionContext jobExecutionContext) throws JobExecutionException {
+    // Must Have , Before Execute the Init, Quartz Requires a Non-Arg Constructor
+    this.init(jobExecutionContext);
+
+    try {
+      List<ChangeEvent> batch = pollEvents(offset, 100);
+      if (!batch.isEmpty()) {
+        boolean success = publishEvents(batch);
+        if (success) {
+          offset += batch.size();
+        } else {
+          handleFailedEvents(batch);
+        }
+      }
+    } catch (InterruptedException e) {
+      LOG.error("Interrupted while polling events", e);
+      Thread.currentThread().interrupt();
+    } catch (Exception e) {
+      handleException(e);
+    } finally {
+      LOG.debug(
+          "Committing offset for eventSubscription {}  {}", eventSubscription.getName(), offset);
+      commitOffset(jobExecutionContext, offset);
```

Comment:
If we are committing the current offset finally, this will be problematic if an exception happens and we need to retry, no?

Reply:
we are handing that , if any events has errors it retried configured number of times then pushed to dlc